Add partial updates loading states for the order summary

ralucaStan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
This PR adds partial loading states to the cart items and prices (total, subtotal, taxes) in the Order summary. These can be seen when:
- a product quantity changes or a new product is added to the cart
- when coupon is added/removed to the cart
- when a shipping method is changed

This PR **does not include** the express area buttons. 
The checkout ones are part of another PR; The cart ones will be handled as part of the work to bring skeletons to Cart

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
To improve the loading behavior with the Skeleton component, the order summary now reads from a single hook that keeps track of changes that impact prices. I opted to create one hook for ease of use and to avoid repetition.

I've also synced the iAPI store actions for changing product quantity with our data store so that when the cart quantities change from an iAPI block, those events are also caught and show loaders in the Order summary.

This PR alse removes the VisuallyHidden component from skeletons which was bringing extra files from the `@woocommerce/components` package.



<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes https://linear.app/a8c/issue/WOOPLUG-3804/cart-and-checkout-blocks-implement-order-summary-partial-updates
Closes https://github.com/woocommerce/woocommerce/issues/57029




<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Make sure you have shipping with prices on your store

#### Add remove/coupon
1. Add a coupon to your cart or checkout blocks that impacts the cart total
2. Notice how the totals show the Skeleton 
3. Notice the updated price when the Skeleton is removed

#### Change shipping method
1. Change between 2 shipping methods with different prices
2. Notice how the totals show the Skeleton 
3. Notice the updated price when the Skeleton is removed

#### Change product quantity in the Cart block
1. Increase the quantity or remove a product from cart
2. Notice how the totals show the Skeleton
3. Notice the updated price when the Skeleton is removed

#### Change product quantity in the Checkout block
1. Add a Product collection block on your checkout page
2. Add a product to page and go to the checkout page
3. Add from the Product collection block a product to cart
4. Notice the skeletons in for prices and for the Cart items

#### Test taxes
Make sure you show taxes with this setup
Notice how the `Including 1,80 € Tax, 19,53 € 18%` text gets a loader as well

#### Test the Mini Cart
Notice that the look of the Cart items hasn't changed in the Mini Cart

For `Display prices during cart and checkout - > Including tax`

<img width="519" height="448" alt="Screenshot 2025-07-10 at 15 57 29" src="https://github.com/user-attachments/assets/4eb9fe16-9bf2-4adb-9eb2-e55820f1c989" />
<img width="496" height="524" alt="Screenshot 2025-07-10 at 15 57 01" src="https://github.com/user-attachments/assets/89b10947-6e35-4839-927b-c8b5642a8713" />

For `Display prices during cart and checkout - > Excluding tax`
there will be subtotal tax items
<img width="285" height="430" alt="Screenshot 2025-07-10 at 16 04 38" src="https://github.com/user-attachments/assets/ccdec5fc-9632-4fda-a712-d85489442000" />


<br class="Apple-interchange-newline">

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ralucaStan committed:
Add partial updates loading states for the order summary

ralucaStan committed:
Merge branch 'trunk' into wooplug-3804-implement-partial-updates-loading-states-for-the-order

ralucaStan committed:
Account for cart item updates coming from the interactivity api blocks
- sync the productsPendingAdd, itemsPendingQuantity, itemsPendingDelete from iapi
- account for these values in cartIsLoading

ralucaStan committed:
remove type comment

ralucaStan committed:
Fix js failing tests

ralucaStan committed:
Add changelog

ralucaStan committed:
fix lint issue

ralucaStan committed:
Fix lint and failing tests

ralucaStan committed:
test resources on the CI

ralucaStan committed:
Merge branch 'trunk' into wooplug-3804-implement-partial-updates-loading-states-for-the-order

ralucaStan committed:
Merge branch 'trunk' into wooplug-3804-implement-partial-updates-loading-states-for-the-order

ralucaStan committed:
Remove the VisuallyHidden component from the skeleton
The components brings unnecessary js files to the build so I've rewritten the skeleton to not use it.

ralucaStan committed:
Merge branch 'wooplug-3804-implement-partial-updates-loading-states-for-the-order' of github.com:woocommerce/woocommerce into wooplug-3804-implement-partial-updates-loading-states-for-the-order

ralucaStan committed:
Fix unit test

ralucaStan committed:
remove hook parameter, it's not needed for reading loading states

ralucaStan committed:
Merge branch 'trunk' into wooplug-3804-implement-partial-updates-loading-states-for-the-order

ralucaStan committed:
Fix the Mini cart cart items styles

ralucaStan committed:
Merge branch 'trunk' into wooplug-3804-implement-partial-updates-loading-states-for-the-order

mikejolley commented on plugins/woocommerce/client/blocks/assets/js/base/context/hooks/use-order-summary-loading-state.ts:
```diff
@@ -0,0 +1,38 @@
+ * Custom hook that provides comprehensive loading states for Order Summary components.
+ * This hook combines various loading states that can affect order totals and pricing.
+ */
+export const useOrderSummaryLoadingState = () => {
```
Do you think it would make sense to include the isLoading value here via another hook such as `useStoreCart` or is this needed because it touches other stores? There is a checkout hook also I guess it could live in.

ralucaStan replied:
I wanted the scope of this hook to be clear. The hook is used in both cart and checkout. What are the disadvantages you see for creating a separate hook?

mikejolley commented on plugins/woocommerce/tests/e2e-pw/tests/js-file-monitor/monitor-js-file-number.spec.js:
```diff
@@ -61,6 +61,20 @@ test.describe( 'Keeps track of the number of JS files included on key shopper pa
+			scriptSources.forEach( ( src, index ) => {
+				console.log( `  ${ index + 1 }. ${ src }` );
+			} );
+			console.log( '' );
```
What was this for?

ralucaStan replied:
I had to debug an e2e test; looks like I missed fully cleaning up this file

mikejolley reviewed (dismissed):
Added a few questions. I like how the totals area reacts to changes. I think if I could improve anything it would be to find a way to preserve sizes of content when changing to loading state. Example, if the skeleton here matched the size of the original tax string:

https://github.com/user-attachments/assets/0cb216b3-27ac-414f-9074-ab2cb34f1d65

I realise that might be out of scope but it would also be a nice enhancement! But generally, the order summary area on checkout looks great, thank you.

There are some issues however on the cart page, and within mini-cart which I believe are related to this PR somehow. 

On the cart page, when I change a quantity, the items vanish and it goes back to the default skeleton, with a loading spinner which is a regression from trunk:

https://github.com/user-attachments/assets/3f885378-8bef-4ec2-a42d-1762b6b94ee8

Similarly, the mini cart items get skeletonized, and the header above disspears:

https://github.com/user-attachments/assets/cc488df8-488b-4db2-b648-ad222b737d22

Could these be related to the isLoading changes?

ralucaStan committed:
Fix useCartStore isCartLoading prop and add a separate hasPendingItemsOperations value

isCartLoading is referring to cart calls, while pending operations are separate.
While both should show loaders, those loaders need to be shown in separate places;
isCartLoading - cart items list
hasPendingItemsOperations  - order summary

ralucaStan committed:
remove debugging code

ralucaStan committed:
Fix Mini cart items styles

ralucaStan committed:
Merge branch 'trunk' into wooplug-3804-implement-partial-updates-loading-states-for-the-order

mikejolley reviewed (commented):
Thanks for fixing the loading issue. I've confirmed it fixes the mini-cart and regular cart. I tried on a small screen and didn't spot any issues in the order confirmation area either.

The only issue I did notice was a small content jump while things are loading due to the size of the skeleton:

https://github.com/user-attachments/assets/bc807eb3-ece0-4cb1-a53c-1b3dd50e486f

Do you think this would be an easy fix?

ralucaStan committed:
Fix content jump for taxes skeleton

ralucaStan committed:
Merge branch 'wooplug-3804-implement-partial-updates-loading-states-for-the-order' of github.com:woocommerce/woocommerce into wooplug-3804-implement-partial-updates-loading-states-for-the-order

straku commented:
> FYI @straku once the font sizes will be fixed we'll need to also update these Skeletons

@ralucaStan I am planning to tackle order summary last. Thanks for the heads-up.

ralucaStan committed:
Fix content jump for tax line item

straku commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/totals/footer-item/index.tsx:
```diff
@@ -140,30 +140,33 @@ const TotalsFooterItem = ( {
+						<p className="wc-block-components-totals-footer-item-tax">
+							{ isLoading ? (
+								<>
+									<span>Including </span>
```
Should we translate the whole text? We risk a situation in which the real message will be in a different language vs the loading state hardcoded to English 🤔

straku reviewed (approved):
I was mostly focused on manually testing the change, since it was already code-reviewed by Mike. I second his opinion that the skeleton should match the size of the content that it is replacing, but I don't see a problem with follow-up changes that would improve that.

I cannot reproduce the issue with the tiny layout shift in cart. In general, the experience is very smooth. One potential improvement is to consider not showing loading state when async operations are very fast—if it takes less than 200ms then just switch to the new value, if it's longer show the loader.

ralucaStan committed:
Merge branch 'trunk' into wooplug-3804-implement-partial-updates-loading-states-for-the-order

ralucaStan committed:
Use translations for placeholder text
